File: practice/bubble.py
def bubbleSort(a, n):
    countCmpOps = 0
    countSwaps = 0

    for p in range(1, n):
        for i in range(1, n - p + 1):
            countCmpOps += 1
            if a[i - 1] > a[i]:
                countSwaps += 1
                a[i - 1], a[i] = a[i], a[i - 1]

    return countCmpOps, countSwaps

def bubbleSortImproved1(a, n):
    countCmpOps = 0
    countSwaps = 0

    for p in range(1, n):
        swapped = False
        for i in range(1, n - p + 1):
            countCmpOps += 1
            if a[i - 1] > a[i]:
                countSwaps += 1
                a[i - 1], a[i] = a[i], a[i - 1]
                swapped = True
        if not swapped:
            break

    return countCmpOps, countSwaps

def bubbleSortImproved2(a, n):
    countCmpOps = 0
    countSwaps = 0
    lastSwappedPos = n

    while lastSwappedPos > 0:
        swappedPos = 0
        for i in range(1, lastSwappedPos):
            countCmpOps += 1
            if a[i - 1] > a[i]:
                countSwaps += 1
                a[i - 1], a[i] = a[i], a[i - 1]
                swappedPos = i
        lastSwappedPos = swappedPos

    return countCmpOps, countSwaps

def main():
    numTestCases = int(input())
    results = []

    for i in range(numTestCases):
        b = list(map(int, input().split()))
        num = b.pop(0)

        cmpOps1, swaps1 = bubbleSort(b.copy(), num)
        cmpOps2, swaps2 = bubbleSortImproved1(b.copy(), num)
        cmpOps3, swaps3 = bubbleSortImproved2(b.copy(), num)

        results.append((cmpOps1, swaps1, cmpOps2, swaps2, cmpOps3, swaps3))

    for cmpOps1, swaps1, cmpOps2, swaps2, cmpOps3, swaps3 in results:
        print(f"{cmpOps1} {swaps1} {cmpOps2} {swaps2} {cmpOps3} {swaps3}")

if __name__ == "__main__":
    main()

File: data_structure/bubble.c
#include <stdio.h>

#define MAX_SIZE 1000
void bubbleSort(int a[], int n);
void bubbleSortImproved1(int a[], int n);
void bubbleSortImproved2(int a[], int n);
void copyArray(int a[], int b[], int n);

void swap(int* a, int* b){
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

void copyArray(int a[], int b[], int n){
    for (int i=0; i<n; i++)
        a[i] = b[i];
}

void bubbleSort(int a[], int n) {
    int countCmpOps = 0;
    int countSwaps = 0;

    for (int p = 1; p < n; p++) {
        for (int i = 1; i < n - p + 1; i++) {
            countCmpOps++;
            if (a[i - 1] > a[i]) {
                countSwaps++;
                swap(&a[i-1], &a[i]);
            }
        }
    }

    printf("%d %d ", countCmpOps, countSwaps);
}

void bubbleSortImproved1(int a[], int n) {
    int countCmpOps = 0;
    int countSwaps = 0;

    for (int p = 1; p < n; p++) {
        int swapped = 0;
        for (int i = 1; i < n - p + 1; i++) {
            countCmpOps++;
            if (a[i - 1] > a[i]) {
                countSwaps++;
                swap(&a[i-1], &a[i]);
                swapped = 1;
            }
        }
        if (!swapped) {
            break;
        }
    }

    printf("%d %d ", countCmpOps, countSwaps);
}

void bubbleSortImproved2(int a[], int n) {
    int countCmpOps = 0;
    int countSwaps = 0;
    int lastSwappedPos = n;

    while (lastSwappedPos > 0) {
        int swappedPos = 0;
        for (int i = 1; i < lastSwappedPos; i++) {
            countCmpOps++;
            if (a[i - 1] > a[i]) {
                countSwaps++;
                swap(&a[i-1], &a[i]);
                swappedPos = i;
            }
        }
        lastSwappedPos = swappedPos;
    }

    printf("%d %d ", countCmpOps, countSwaps);
}

int main() {
    int numTestCases;
    scanf("%d", &numTestCases);

    for (int i = 0; i < numTestCases; ++i) {
        int num;
        int a[MAX_SIZE], b[MAX_SIZE];

        scanf("%d", &num);

        for (int j = 0; j < num; j++) {
            scanf("%d", &b[j]);
        }

        copyArray(a, b, num);
        bubbleSort(a, num);

        copyArray(a, b, num);
        bubbleSortImproved1(a, num);

        copyArray(a, b, num);
        bubbleSortImproved2(a, num);
        printf("\n");
    }

    return 0;
}